toc/rfc/rfc-draft-add-support-for-file-based-service-binding.md

```diff
@@ -0,0 +1,111 @@
+# Meta
+[meta]: #meta
+- Name: Add Support for File based Service Binding Information
+- Start Date: 2024-03-11
+- Author(s): @beyhan
+- Contributing Authors: @stephanme, @loewenstein
+- Status: Draft
+- RFC Pull Request: (fill in with PR link after you submit it)
+
+
+
+## Summary
+
+The current contract between the CF platform and applications for service binding information is based on an environment variable. The Linux Kernel defines size limit per environment variable and there are also other limitations with this approach. That is why, CF should add support for an alternative option to provide service binding information which can address the limitations of the current approach.
+
+## Problem
+
+The CF platform provides Service Binding Information to hosted applications via the [VCAP_SERVICES environment variable](https://docs.cloudfoundry.org/devguide/services/application-binding.html). There are following challenges with this approach:
+- The environment variable length has a hard limit of `131072` bytes per variable which is controlled by the underlying [Linux kernel](https://github.com/torvalds/linux/blob/master/include/uapi/linux/binfmts.h). The environment variable size is defined by `MAX_ARG_STRLEN`, which is a constant with the value `PAGE_SIZE*32` where page size is `4096` bytes. This means that to change the size limit for the CF platform a recompiled kernel with updated value for `MAX_ARG_STRLEN` is required. This limit could be an issue for applications using many services. If the limit is reached by an application, it will fail to stage as discussed in [this issue](https://github.com/cloudfoundry/garden-runc-release/issues/160).
+- Updates of the Service Binding Information require restage. This is not optimal for an eventual support of [Binding rotation](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#binding-rotation) specification from the OSBI API spec.
+
+
+## Proposal
+
+The CFF community should implement an alternative approach for service binding information based on `tmpfs` file(s). Using a file or files to provide service binding information to applications will address the challenges listed in the problem section because:
```

Comment:
While being generally very supportive of this idea, it's important to consider the permissions of this file. Currently, when integrated with CredHub, Diego looks up service credentials just before starting an App and injects them into the App's environment, so that only the App (and not anyone connecting via `cf ssh`) can see the credentials in plaintext. If plaintext credentials are moved to a file, then it's important to think about who can see this file. Is it just the App, or should `cf ssh`, `cf run-task`, etc... see them? On my system both Apps and `cf ssh` run with `uid=2000(vcap)`, so any files that an App can read are also readable by `cf ssh` or `cf run-task` users.

Reply:
Thanks @blgm for this important point and I also agree that these file should bring the same security level as we have today with `VCAP_SERVICES`. I wonder about:
>Currently, when integrated with CredHub, Diego looks up service credentials just before starting an App and injects them into the App's environment, so that only the App (and not anyone connecting via cf ssh) can see the credentials in plaintext.

When I do `cf ssh` into an app and execute the command `env` I can see the content of the `VCAP_SERVICES` env var or any env var Diego configured for the app container. With that I can see the credentials which are in plain text in `VCAP_SERVICES` services with `cf ssh` already today or do I miss something here?

Reply:
https://github.com/cloudfoundry/buildpackapplifecycle/blob/main/credhub/credhub.go#L26-L45
At least for buildpack based apps this seems to be controlled by the `CREDHUB_SKIP_INTERPOLATION` environment variable.

Reply:
I haven't tried in a while, but I'm guessing you can still `cat /proc/<pid>/environ [see](https://www.baeldung.com/linux/process-environment-variables) to view the values, even as careful as CredHub/Diego are, I'm pretty sure you can still see those values.

+1 for secure file permissions, probably `600` owned by the uid running the app.

I don't think you can block someone sufficiently motivated from seeing them if they have `cf ssh` access though.

Reply:
Yes, it's definitely resolved from my perspective.